File: src/ipc/watchlist.ts
import { invoke } from "@tauri-apps/api/core";
import { Movie } from "../types";

export async function addToWatchlist(movieId: string): Promise<void> {
  await invoke("add_to_watchlist", { movieId });
}

export async function removeFromWatchlist(movieId: string): Promise<void> {
  await invoke("remove_from_watchlist", { movieId });
}

export async function isInWatchlist(movieId: string): Promise<boolean> {
  return await invoke<boolean>("is_in_watchlist", { movieId });
}

export async function listWatchlist(): Promise<Movie[]> {
  return await invoke<Movie[]>("list_watchlist");
}

File: src-tauri/src/commands/watchlist.rs
use std::sync::Arc;
use tauri::State;
use uuid::Uuid;

use crate::domain::Movie;
use crate::error::AppError;
use crate::services::WatchlistService;

#[tauri::command]
pub async fn add_to_watchlist(
    movie_id: String,
    watchlist_service: State<'_, Arc<WatchlistService>>,
) -> Result<(), AppError> {
    let id = Uuid::parse_str(&movie_id).map_err(|e| AppError::Validation(e.to_string()))?;
    watchlist_service.add(&id)
}

#[tauri::command]
pub async fn remove_from_watchlist(
    movie_id: String,
    watchlist_service: State<'_, Arc<WatchlistService>>,
) -> Result<(), AppError> {
    let id = Uuid::parse_str(&movie_id).map_err(|e| AppError::Validation(e.to_string()))?;
    watchlist_service.remove(&id)
}

#[tauri::command]
pub async fn is_in_watchlist(
    movie_id: String,
    watchlist_service: State<'_, Arc<WatchlistService>>,
) -> Result<bool, AppError> {
    let id = Uuid::parse_str(&movie_id).map_err(|e| AppError::Validation(e.to_string()))?;
    watchlist_service.is_in_watchlist(&id)
}

#[tauri::command]
pub async fn list_watchlist(
    watchlist_service: State<'_, Arc<WatchlistService>>,
) -> Result<Vec<Movie>, AppError> {
    watchlist_service.list()
}
